pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -382,23 +595,55 @@ pub mod pallet {
         fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
             let mut weight = Weight::zero();
 
-            // 1 write in on_finalize
-            weight += T::DbWeight::get().writes(1);
+            weight += T::DbWeight::get().writes((2 * T::MaxInFlightOrders::get() + 3) as u64);
+            weight += T::DbWeight::get().reads(3);
 
             weight
         }
 
-        fn on_finalize(_: BlockNumberFor<T>) {
-            // We clear this storage item because we only need it to prevent collators from buying
-            // more than one core for the same parathread in the same relay block
-            // TODO: with async backing, it is possible that two tanssi blocks are included in the
-            // same relay block, so this kill is not correct, should only kill the storage if the
-            // relay block number has changed
-            // TODO: this allows collators to send N consecutive messages to buy 1 core for the same
-            // parathread, as long as the parathread block is not included in pallet_author_noting.
-            // So a malicious collator could drain the parathread tank account by buying cores on
-            // every tanssi block but not actually producing any block.
-            InFlightOrders::<T>::kill();
+        fn on_finalize(n: BlockNumberFor<T>) {
+            let mut ttl_queue = InFlightOrdersTtl::<T>::get();
```

Comment:
if we want to do stuff on_finalize, we have to take the weight into account on_initialize

Reply:
ah sorry we are counting it :+1: